### ISSUE TITLE
Assertion failed: `has_breaking_edge` while computing matchtigs

### ISSUE THREAD

**Issue by enricorox** [algbio/matchtigs#6]
Hi!
Sorry for the inconvenience, I think I found a bug.
I was computing matchtigs for the sequence SRR332538_1, k=21.
The program failed to check the assertion `has_breaking_edge`. 
I got the warning `11:19:03 [WARN] Inserted 12 bidirected loops` that may be related.

> 11:09:34 [INFO] Logging initialised successfully
> 11:09:34 [INFO] Reading bcalm2 fa as edge centric bigraph with k = 21 from "SRR332538_1.fasta-nc.unitigs.fa"
> 11:09:36 [INFO] Loading took 2.1 seconds
> 11:09:36 [INFO] The sequences take a total 6MiB of memory
> 11:09:36 [INFO] k = 21
> 11:09:36 [INFO] Graph has 1353449 nodes and 1440668 edges
> 11:09:36 [INFO] Computing edge weights for shortest path queries
> 11:09:36 [INFO] Computing matchtigs
> 11:09:36 [INFO] Collecting nodes with missing incoming or outgoing edges
> 11:09:36 [INFO] Found 651311 nodes with missing outgoing edges
> 11:09:36 [INFO] Found 651311 nodes with missing incoming edges
> 11:09:36 [INFO] Of those there are 72 self-mirrors
> 11:09:36 [INFO] Computing shortest paths between nodes with missing outgoing and nodes with missing incoming edges
> 11:09:36 [INFO] Using 16 threads to run ~651311 dijkstras
> 11:09:37 [INFO] Creating dijkstra threads
> 11:09:37 [INFO] Waiting for dijkstra threads to finish
> 11:09:40 [INFO] Found 22481342 shortest paths
> 11:09:47 [INFO] Took 10.943342s for computing paths and getting edges, of this 3.431739s are from dijkstra
> 11:09:47 [INFO] Found 263187 nodes and 18428066 edges
> 11:09:47 [INFO] Matching problem contains 173 edges that originate from 488 mirror biedges
> 11:09:47 [INFO] Computing WCCs of graph
> 11:09:48 [INFO] Filtering irrelevant WCCs
> 11:09:48 [INFO] Found 1773 relevant WCCs
> 11:09:48 [INFO] Assigning extra nodes to nodes in the matching instance
> 11:09:48 [INFO] Outputting matching problem to "SRR332538_1.fasta.matchtigs.fa.minimalperfectmatching"
> 11:09:56 [INFO] Running matcher at "blossom5"
> perfect matching with 533466 nodes and 38172067 edges
>     starting init...done [529.412 secs]. 490 trees
>     .256.128.64.56.48.40.32.24.16.8.6.4.2.0.
> done [530.213 secs]. 29054 grows, 0 expands, 970 shrinks
>     expands: [0.000 secs], shrinks: [0.001 secs], dual updates: [0.000 secs]
> cost = 3429126.0
> 11:19:03 [INFO] Applying matcher result to graph
> 11:19:03 [INFO] Inserted 122842 matched edges
> 11:19:03 [WARN] Inserted 12 bidirected loops
> 11:19:03 [INFO] Making graph Eulerian by completing unmatched nodes
> 11:19:03 [INFO] Adding edges for 546248 unmatched in_nodes, 546248 unmatched out_nodes and 42 unmatched self_mirror_nodes
> 11:19:03 [INFO] 287490 edges need to be added in total
> 11:19:03 [INFO] 523045/17716/5487/0 binodes that are not self-mirrors have difference 1/2/3/4
> 11:19:04 [INFO] Finding Eulerian bicycle
> 11:19:04 [INFO] Found 1277 Eulerian bicycles
> 11:19:04 [INFO] Breaking Eulerian bicycles at expensive temporary edges
> 11:19:04 [INFO] Processing Eulerian bicycle with 1064152 biedges
> 11:19:04 [INFO] Processing Eulerian bicycle with 5 biedges
> 11:19:04 [INFO] Processing Eulerian bicycle with 1 biedges
> thread 'main' panicked at 'assertion failed: has_breaking_edge', src/implementation/matchtigs/mod.rs:895:9
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
>    1: core::panicking::panic_fmt
>              at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
>    2: core::panicking::panic
>              at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:48:5
>    3: matchtigs::implementation::matchtigs::compute_matchtigs
>              at /opt/conda/conda-bld/matchtigs_1670488620267/work/src/implementation/matchtigs/mod.rs:895:9
>    4: matchtigs::implementation::matchtigs::compute_matchtigs_choose_node_weight_array_type
>              at /opt/conda/conda-bld/matchtigs_1670488620267/work/src/implementation/matchtigs/mod.rs:127:13
>    5: matchtigs::implementation::matchtigs::compute_matchtigs_choose_heap_type
>              at /opt/conda/conda-bld/matchtigs_1670488620267/work/src/implementation/matchtigs/mod.rs:93:13
>    6: <matchtigs::implementation::matchtigs::MatchtigAlgorithm<SequenceHandle> as matchtigs::implementation::TigAlgorithm<Graph>>::compute_tigs
>              at /opt/conda/conda-bld/matchtigs_1670488620267/work/src/implementation/matchtigs/mod.rs:70:9
>    7: matchtigs::main
>              at /opt/conda/conda-bld/matchtigs_1670488620267/work/src/bin.rs:1135:25
>    8: core::ops::function::FnOnce::call_once
>              at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:248:5

**Comment by sebschmi** [comment]
Thanks, I will take a look!

**Comment by sebschmi** [comment]
This seems to be a read dataset. Which min-abundance are you using? The default 2 of bcalm?

**Comment by enricorox** [comment]
No, I'm using min-abundance 1

Il mar 27 giu 2023, 08:21 Sebastian Schmidt ***@***.***> ha
scritto:

**Comment by sebschmi** [comment]
Okay, that is strange, because when I run this dataset through bcalm with min-abundance 1, I get a graph with 9 million nodes, and you get some 1.3 million. Am I doing something different when unpacking the SRR file? I am using `fastq-dump --stdout --fasta default <input> > <output>`

**Comment by sebschmi** [comment]
If you would be able to provide your file `SRR332538_1.fasta-nc.unitigs.fa`, that would be great!

**Comment by enricorox** [comment]
I had split the read dataset in two paired-end with `fasterq-dump --fasta --split-files SRR332538.sra` and used the first one. However, the bcalm file is 
[SRR332538_1.fasta-nc.unitigs.zip](https://github.com/algbio/matchtigs/files/11878483/SRR332538_1.fasta-nc.unitigs.zip)

Thanks!

**Comment by sebschmi** [comment]
Thanks for the file! I found the issue: the assertion did not account for the fact that some components of the graph may be Eulerian already, and still wanted their Eulerian cycles to contain breaking arcs. I have weakened the assertion, and now your test case works. Thank you for the report!

The new version 2.1.4 contains the fix. It will take some time until it arrives on bioconda, because of this issue: https://github.com/bioconda/bioconda-recipes/issues/41526
